src/prefect/_versioning.py

```diff
@@ -21,82 +21,370 @@ class SimpleVersionInfo(VersionInfo):
 class GithubVersionInfo(VersionInfo):
     type: Literal["vcs:github"] = "vcs:github"
     version: str
+    commit_sha: str
+    message: str
     branch: str
-    url: str
     repository: str
+    url: str
 
     @model_validator(mode="after")
     def validate_branch(self):
         if not self.branch:
             raise ValueError("branch is required when type is 'vcs:github'")
         return self
 
+    @model_validator(mode="after")
+    def validate_commit_sha(self):
+        if not self.commit_sha:
+            raise ValueError("commit_sha is required when type is 'vcs:github'")
+        return self
+
+
+class GitlabVersionInfo(VersionInfo):
+    type: Literal["vcs:gitlab"] = "vcs:gitlab"
+    version: str
+    commit_sha: str
+    message: str
+    branch: str
+    repository: str
+    url: str
+
+    @model_validator(mode="after")
+    def validate_branch(self):
+        if not self.branch:
+            raise ValueError("branch is required when type is 'vcs:gitlab'")
+        return self
+
+    @model_validator(mode="after")
+    def validate_commit_sha(self):
+        if not self.commit_sha:
+            raise ValueError("commit_sha is required when type is 'vcs:gitlab'")
+        return self
```

Comment:
i think we don't need `model_validator` here if we're just checking that a field is provided, unless its to provide specific error messages, in which case I think we should only need one per subclass